pkg/reconciler/clusterducktype/clusterducktype.go

```diff
@@ -18,37 +18,127 @@ package clusterducktype
-func (r *Reconciler) ReconcileKind(ctx context.Context, o *discoveryv1alpha1.ClusterDuckType) reconciler.Event {
-	logger := logging.FromContext(ctx)
+// ReconcileKind implements Interface
+func (r *Reconciler) ReconcileKind(ctx context.Context, dt *v1alpha1.ClusterDuckType) reconciler.Event {
+	/// By query
 
-	crd, err := r.crdLister.Get("channels.messaging.knative.dev") // TODO generalize past testing
-	if err != nil {
-		logger.Errorf("error getting crd: %q", err)
-	} else {
-		logger.Errorf("found crd: %q", crd)
+	kinds := make(map[string]*apiextensionsv1.CustomResourceDefinition, 0)
+
+	for _, st := range dt.Spec.Selectors {
+		crds, err := r.getCRDsWith(st.LabelSelector)
+		if err != nil {
+			// TODO: this should be a condition that reports back that RBAC is incorrect for getting CRDs?
+			return err
+		}
+		for _, crd := range crds {
+			key := crd.Name
+			if _, found := kinds[key]; !found {
+				kinds[key] = crd
+			}
+		}
+	}
+
+	foundDucks := make([]v1alpha1.FoundDuck, 0)
+	for _, crd := range kinds {
+		foundDucks = append(foundDucks, CRDToFoundDuck("", crd))
+	}
+
+	/// By ref
+
+	for _, dv := range dt.Spec.Versions {
+		for _, ref := range dv.Refs {
+			// TODO we should query and test that the Ref is installed and works on this cluster.
+			foundDucks = append(foundDucks, v1alpha1.FoundDuck{
+				DuckVersion: dv.Name,
+				Ref:         ref,
+			})
+		}
 	}
 
-	// TODO: remove this when there is real logic here.
-	o.Status.MarkReady()
+	// TODO: the above logic needs to pop out the label selected found ducks to
+	// each version supported that is not overwritten. This will likely be
+	// slightly complicated and needs a bunch of tests. Break it apart into
+	// some smaller component that we can encapsulate those tests.
 
+	// Sort and store.
+
+	sort.Sort(ByFoundDuck(foundDucks))
+	dt.Status.DuckList = foundDucks
+	dt.Status.DuckCount = DuckCount(foundDucks)
+
+	dt.Status.MarkReady()
 	return nil
 }
+
+// getCRDsWith returns CRDs labeled as given.
+// labelSelector should be in the form "duck.knative.dev/source=true"
+func (r *Reconciler) getCRDsWith(labelSelector string) ([]*apiextensionsv1.CustomResourceDefinition, error) {
+	ls, err := labels.Parse(labelSelector)
+	if err != nil {
+		return nil, err
+	}
+
+	list, err := r.crdLister.List(ls)
+	if err != nil {
+		return nil, err
+	}
+
+	return list, nil
+}
+
+// ByFoundDuck implements sort.Interface for []v1alpha1.FoundDuck based on
+// the group and resource fields.
+type ByFoundDuck []v1alpha1.FoundDuck
+
+func (a ByFoundDuck) Len() int      { return len(a) }
+func (a ByFoundDuck) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+func (a ByFoundDuck) Less(i, j int) bool {
+	// TODO: this needs to sort like this, but also group by duck version.
+	keyI := a[i].Ref.Group + a[i].Ref.Resource
+	keyJ := a[j].Ref.Group + a[j].Ref.Resource
+	return keyI < keyJ
+}
+
+func CRDToFoundDuck(duckVersion string, crd *apiextensionsv1.CustomResourceDefinition) v1alpha1.FoundDuck {
```

Comment:
return (*v1alpha1.FoundDuck, error) maybe?
Are you doing something with the empty FoundDuck?

Reply:
or just a pointer? and maybe nil at the end?

Reply:
yeah, good catch, I added a todo to break this logic out into a separate area to test it heavily which I will do next.

Reply:
ack